src/main/java/jenkins/branch/SpecificDescriptor.java

```diff
@@ -0,0 +1,102 @@
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package jenkins.branch;
+
+import hudson.Extension;
+import hudson.ExtensionList;
+import hudson.init.InitMilestone;
+import hudson.init.Initializer;
+import hudson.model.Descriptor;
+import hudson.model.DescriptorVisibilityFilter;
+import hudson.model.ItemGroup;
+import hudson.model.TopLevelItem;
+import hudson.model.TopLevelItemDescriptor;
+import hudson.model.View;
+import jenkins.scm.api.SCMNavigatorDescriptor;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+
+/**
+ * Ensures that when we have a non-null {@link SCMNavigatorDescriptor#newInstance(String)}, a special <b>New Item</b> entry is displayed instead of the generic {@link OrganizationFolder.DescriptorImpl}.
+ */
+@SuppressWarnings("rawtypes") // not our fault
+@Restricted(NoExternalUse.class)
+public class SpecificDescriptor extends TopLevelItemDescriptor {
+
+    public final SCMNavigatorDescriptor delegate;
+
+    SpecificDescriptor(SCMNavigatorDescriptor delegate) {
+        super(TopLevelItem.class); // do not register as OrganizationFolder
+        this.delegate = delegate;
+    }
+    
+    @Override
+    public String getId() {
+        return OrganizationFolder.class.getName() + "." + delegate.getId(); // must be distinct from OrganizationFolder.DescriptorImpl
+    }
+    
+    @Override
+    public String getDisplayName() {
+        return delegate.getDisplayName();
+    }
+
+    @Override
+    public TopLevelItem newInstance(ItemGroup parent, String name) {
+        OrganizationFolder p = new OrganizationFolder(parent, name);
+        p.getNavigators().add(delegate.newInstance(name));
+        return p;
+    }
+
+    @SuppressWarnings("deprecation") // dynamic registration intentional here
+    @Initializer(after=InitMilestone.PLUGINS_STARTED, before=InitMilestone.EXTENSIONS_AUGMENTED)
+    public static void addSpecificDescriptors() {
```

Comment:
Not one I am happy with. I was thinking for a while we needed a way to inject additional items into an `ExtensionList`, but in this case I think it would suffice for `DescriptorVisibilityFilter` to be permitted to _add_ descriptors as well as remove them, since this is purely a UI modification.